servo-tomcat/src/main/java/com/netflix/servo/publish/tomcat/TomcatPoller.java

```diff
@@ -0,0 +1,159 @@
+
+    private static Tag catalina = Tags.newTag("source", "Catalina");
+
+    private static Metric toMetric(long t, ObjectName name, Attribute attribute, Tag dsType) {
+        Tag pool = Tags.newTag("pool", name.getKeyProperty("name"));
+        Tag clazz = Tags.newTag("class", name.getKeyProperty("type"));
+        TagList list = BasicTagList.of(catalina, pool, clazz, dsType);
+        return new Metric(attribute.getName(), list, t, attribute.getValue());
+    }
+
+    private static Metric toGauge(long t, ObjectName name, Attribute attribute) {
+        return toMetric(t, name, attribute, DataSourceType.GAUGE);
+    }
+
+    private static Metric toCounter(long t, ObjectName name, Attribute attribute) {
+        return toMetric(t, name, attribute, DataSourceType.COUNTER);
+    }
+
+    private static final String[] THREAD_POOL_ATTRS = new String[]{
+            "maxThreads",
+            "currentThreadCount",
+            "currentThreadsBusy",
+            "backlog"
+    };
+
+    private static final String[] GLOBAL_REQ_ATTRS = new String[]{
+            "requestCount",
+            "errorCount",
+            "bytesSent",
+            "processingTime",
+            "maxTime"
+    };
+
+    private static final String[] EXECUTOR_ATTRS = new String[]{
+            "poolSize",
+            "activeCount"
+    };
+
+    private static void fetchRequestProcessorMetrics(long now, MBeanServer mbs,
+                                                     List<Metric> metrics) throws JMException {
+        final ObjectName globalName = new ObjectName("Catalina:type=GlobalRequestProcessor,*");
+        final Set<ObjectName> names = mbs.queryNames(globalName, null);
+        if (names == null) {
+            return;
+        }
+
+        for (ObjectName name : names) {
+            AttributeList list = mbs.getAttributes(name, GLOBAL_REQ_ATTRS);
+            for (Attribute a : list.asList()) {
+                if (!a.getName().equals("maxTime")) {
+                    metrics.add(toCounter(now, name, a));
+                } else {
+                    // the only gauge here is maxTime
+                    metrics.add(toGauge(now, name, a));
+                }
+            }
+        }
+        LOGGER.info("GlobalReqProc={}", metrics);
+    }
+
+    private static void fetchThreadPoolMetrics(long now, MBeanServer mbs, List<Metric> metrics)
+            throws JMException {
+        final ObjectName threadPoolName = new ObjectName("Catalina:type=ThreadPool,*");
+        final Set<ObjectName> names = mbs.queryNames(threadPoolName, null);
+        if (names == null) {
+            return;
+        }
+
+        for (ObjectName name : names) {
+            AttributeList list = mbs.getAttributes(name, THREAD_POOL_ATTRS);
+            for (Attribute a : list.asList()) {
+                metrics.add(toGauge(now, name, a));
+            }
+        }
+    }
+
+    private static void fetchExecutorMetrics(long now, MBeanServer mbs, List<Metric> metrics)
```

Comment:
One issue we have seen with using an executor is that the values for the default ThreadPool mbean can be bogus (-1 for many values). It would be nice if we could make that cleaner. Maybe we shouldn't report the ThreadPool stats if a custom executor is used. I think the `executorName` on the Connector mbean will tell you which one is being used for a given connector.

Another area we have had some confusion in the past is that some of the naming is different when you switch to a custom executor. Should we normalize that a bit so the general names are the same? For example, `currentThreadCount` vs `activeCount`.